tests/tests_launch/test_launch_jobs.py

```diff
@@ -0,0 +1,231 @@
+import json
+import os
+import tempfile
+from unittest import mock
+
+import pytest
+import wandb
+from wandb.errors import CommError, LaunchError
+from wandb.sdk.data_types._dtypes import TypeRegistry
+import wandb.sdk.launch.launch as launch
+from wandb.sdk.launch.launch_add import launch_add
+import wandb.sdk.launch._project_spec as _project_spec
+
+from .test_launch import (
+    check_mock_run_info,
+    code_download_func,
+    EMPTY_BACKEND_CONFIG,
+    mock_load_backend,
+    mocked_fetchable_git_repo,
+)
+
+from ..utils import fixture_open
+
+input_types = TypeRegistry.type_of(
+    {"epochs": 2, "heavy": False, "sleep_every": 0}
+).to_json()
+output_types = TypeRegistry.type_of({"loss": 0.2, "cool": True}).to_json()
+
+
+@pytest.fixture
+def mocked_public_artifact(monkeypatch):
+    def mock_artifact_fetcher(job_download_func):
+        def artifact_fetcher(client, name, type):
+            if type == "job":
+                job_artifact = mock.MagicMock()
+                job_artifact.type = "job"
+                job_artifact.download = job_download_func
+                job_artifact.digest = "job123"
+                return job_artifact
+            else:
+                code_artifact = mock.MagicMock()
+                code_artifact.type = "code"
+                code_artifact.download = code_download_func
+                code_artifact.digest = "code123"
+                return code_artifact
+
+        monkeypatch.setattr(
+            wandb.apis.public.Api,
+            "artifact",
+            lambda *arg, **kwargs: artifact_fetcher(*arg, **kwargs),
+        )
+        monkeypatch.setattr(
+            wandb.sdk.launch._project_spec.wandb.apis.public.Api,
+            "artifact",
+            lambda *arg, **kwargs: artifact_fetcher(*arg, **kwargs),
+        )
+
+    return mock_artifact_fetcher
+
+
+def test_fetch_job_fail(api):
+
+    launch_project = _project_spec.LaunchProject(
+        None,
```

Comment:
This is where the explicit kwarg pattern would be nice. I have no idea what all these different things are.